LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DisposableExtensions.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaWan.NetworkServer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    internal static class DisposableExtensions
+    {
+        /// <remarks>
+        /// In general <see cref="IAsyncDisposable.DisposeAsync"/> implementations are not
+        /// expected to throw exceptions. If any of the disposable objects throw exceptions then
+        /// the behaviour of this method is undefined.
+        /// </remarks>
+        public static async ValueTask DisposeAllAsync(this IEnumerable<IAsyncDisposable> disposables, int concurrency)
+        {
+            ArgumentNullException.ThrowIfNull(disposables, nameof(disposables));
+            if (concurrency <= 0) throw new ArgumentOutOfRangeException(nameof(concurrency), concurrency, null);
+
+            var capacity = disposables switch
+            {
+                ICollection<IAsyncDisposable> collection => collection.Count,
+                IReadOnlyCollection<IAsyncDisposable> collection => collection.Count,
+                _ => (int?)null,
+            };
+
+            if (capacity is 0)
+                return;
+
+            using var semaphore = new SemaphoreSlim(concurrency);
+
+            var tasks = capacity is { } someCapacity ? new List<Task>(someCapacity) : new List<Task>();
```

Comment:
Why is this initial capacity needed here?

Reply:
It's an optimisation to avoid having the list to grow its internal storage dynamically as more elements/tasks are added to it. If we can tell it the initial capacity in advance and it can allocate the right-sized array (list's backing storage) from the start.